Playlist object

wwoast opened the pull request:
A new playlist object is created, along with shuffle methods that manipulate the tracks in this playlist object. The main player has shuffling hooks added, and a shuffle button is added to the Gapless GUI.

Shuffling code still has a couple minor bugs related to playback failing after a shuffle event occasionally, or un-shuffling not landing you on the "natural next-track" you'd expect every time.

regosen commented:
Yup, I noticed the bugs you mentioned while testing this out.

I also noticed that if I switch in/out of shuffle mid-playlist, the track indices don't change, so going back to the previous track with play something different (as expected) but the display will show the same track index (unexpected).

I think it would be more intuitive if switching in/out of shuffle affected the tracklist somehow.  Thoughts?

regosen commented:
Here's an example progression through a playlist with the issue I mentioned above:

(user starts player in Regular Mode)
Track 1/5: song A

(user skips to next track)
Track 2/5: song B

(user skips to next track)
Track 3/5: song C

(user switches to Shuffle Mode)
Track 3/5: song C

(user skips to previous track)
Track 2/5: song E

(user is confused because song B was listed as track 2/5, and now song E is listed as track 2/5).

wwoast commented:
I agree with you -- the track index is the one visually-identifying way a song is named in the GUI, so this should be fixed. This will likely require changes both in the Gapless5FileList and the Gapless5 objects.

I'll try and determine the cause of those shuffle issues too, but they may prove elusive...

wwoast commented:
Yeah, this is taking some logic changes that will require deeper testing, but it will be worth it to have consistent, correct track indexes that aren't required to be part of the incoming JSON object. In fact, the only required members of the array of JSON for shuffle mode are [{ "file": "path-to-file.mp3" }].

wwoast commented:
:( The file list object needs to communicate to the player object what the transformed next/previous index is for a new playlist upon shuffle/deshuffle. Or I could just refactor stuff so that the player's active track index is actually stored in the file list object, and not managed in the Player object. Option #2 seems considerably better.

wwoast commented:
OK, there are still some problems with displaying the proper track info after slipping into shuffle mode, but nexttrack/prevtrack always go into the correct trackId. Just need to debug the things that run on updateDisplay and make sure they're using correct index() [for location in a playlist array] or dispIndex() [for whatever the track playing happens to be, even if the underlying array has changed].

I find myself vaguely wishing to rewrite some of this in terms of events / onVariableChange type stuff, but I don't want to pay the cost of doing a framework just to get this benefit.

wwoast commented:
So the GUI track indexes and times finally work as expected. I also added logic for the back button in shuffle mode, so that it still hits the same song even after the playlist is shuffled. Rego, I'm not totally happy with my code quality here, but at least it's ready for deeper review :)

My tests have uncovered the following remaining issues:
* Multiple next/prev track increments put the player in a "non-playing" state, requiring play to be pushed again. This is a problem in the default player too, but it's easier to hit during shuffle-reload-the-world events.
* Switch to shuffle requires reloading all songs, and Gapless tries to load them all at the same time. At the very least, to prevent servers from throttling the next song download, new song loads should be one at a time, starting new ones only after the previous one finishes. Since this idea really is best for lists of songs versus short clips, I propose the allowed number of concurrent song downloads be a configuration setting.

wwoast commented:
Just finished GUI updates to my own interface for Gapless. Very satisfying to see my original jukebox idea working with both Gapless and shuffle. Rego, unless you see problems or bugs I need to fix, I consider this feature branch finished, and am moving on to issues related to song-downloading and playback techniques on mobile.

I'll do my very best to make upcoming PRs easier to review and merge. :) This feature is just eyebleed, but I did the best I could!

regosen commented:
Thanks for the pull request!  I just tried using it locally with a few test cases, so far so good.  I'll be testing it a lot more over the next day or two, and I'll keep you posted.

regosen commented:
So I tried using your new .js/css files on the demonstration page but I get this in the console log:
```
addIndices — gapless5.js:356 TypeError: inputList.slice is not a function. (In 'inputList.slice()', 'inputList.slice' is undefined)
```
And the first player is stuck on "loading...":

<img width="300" alt="screen shot 2016-04-18 at 9 12 51 pm" src="https://cloud.githubusercontent.com/assets/4589394/14627559/6b6a4e58-05aa-11e6-9f55-6aac3332f0aa.png">

I also tried with another demo (one that only has a single player on the page) and got the same stuck "loading..." page, with these errors:

```
gapless5.js:654 Uncaught TypeError: Cannot read property 'getLength' of undefined
https://ccrma.stanford.edu/~regosen/zenfingerpainting/nerdpop/undefined Failed to load resource: the server responded with a status of 404 (Not Found)
undefined:1 Uncaught (in promise) DOMException: Unable to decode audio data
```

It seems like your version isn't compatible with previous use cases.  Maybe it's better to keep your version on a separate repository?

wwoast commented:
I think the problem is that my test coverage didn't go far enough outside my own jukebox's use case. IMO this is worth getting right for existing Gapless folks -- so I'll clone your test pages (the ZFP and the drum/synth one), draw out a test plan, and get these issues knocked out.

Internally, the FileList object wants to treat the playlist as a JSON, not like an array as the original Player/Source objects does. Fixing your test pages without changing them is a matter of the Player init code better converting arrays into the JSON structure the FileList object consumes.

wwoast commented:
Array tracklist and single-item tracklists now work, but not flawlessly. The single-track jukeboxes index from zero instead of one.

Insert Track doesn't seem to work either, which makes me skeptical the Add/Remove/Insert Track functions work properly outside of the FileList init routines. Looking at the demo page more closely gives me ideas on how to create a testing page that fully covers all the Player / FileList methods.

wwoast commented:
The last commit is annoying. By creating addInitialTrack(), I'm paving the way for addTrack/insertTrack/removeTrack to manipulate the FileList object and work interactively again. It's a kludge, but was the least-invasive way I saw to fix these functions.

Longer term, I think it makes sense to move all track selection and manipulation into the FileList object, including the loadingQueue and sources[] arrays that manage XHRs. My hope is that the FileList object will totally contain the playlist state, making it easier to expose portions of this state to other objects, like an XHR request manager that better manages AudioBuffer usage and download rates.

wwoast commented:
Things have been progressing well, now that I've cloned a couple Gapless public sites and made them my test pages, alongside my own jukebox. I believe Shuffle is working correctly now -- though for small playlists I need to add code that disables the button. Learned some hard lessons about shallow and deep copies in the process!

The major remaining bug I know of is that player.remove() futzes track indexing and playlist size tracking. Fingers crossed I can get that figured out this weekend.

wwoast commented:
OK, I've nailed the extant bugs I was tracking. You can now add and remove tracks without affecting the active playback, regardless of shuffle mode. Also, for only 1-2 track lists, the shuffle mode starts grayed out -- however, if you add tracks later, it stays permanently enabled, the assumption being add/remove track events can happen continuously.

regosen commented:
Whoa, much better!  Here's a version of the demonstration page using your branch, it seems to work perfectly:

https://ccrma.stanford.edu/~regosen/gapless5/wwoast.html

I've also switched the Zen Finger Painting album page to use it as well, but there's a couple issues with it:

http://www.zenfingerpainting.com

1. The autostart is broken.  Previously it would automatically start playing track 1 when finished loading.
2. When I shuffle and play the "next" track, I get a callback but I don't know which track is next.  As a result, the visuals on my site no longer match the player.  Also if I click a track while in shuffle mode, it no longer corresponds to what is being played.

I realize issue 2 is a very different use-case from what you intended, and in this case I don't expect you to accommodate it.  Instead, is there a way for me to just hide the shuffle button from the player?  That way, pages like this that are really not geared for shuffle can still work with your changes.

wwoast commented:
I think "hiding shuffle" is a good feature suggestion. However I'm eager to fix issues because the vagaries of shuffle and the new playlist object have required totally revamping how song indexes work internally. I'll see if there's a way to fix ZFP without changes, but I suspect the ZFP callback functions may require minor changes to support shuffle. For that I'm happy to give you a patch. 

Given how this PR may break other cases than ZFP, I think Gapless will require a version bump, and documentation about how to upgrade (an outline of the changes to fix ZFP should fit the bill). Instead of using sources[], the FileList object has a "get()" method that can be used to determine the current playing song, regardless of mode.

BTW, I did not notice autostart failures with my cloned ZFP test page -- I'm on slow Internet at the moment and the song just took a little while to start. I'll dig on that too.

wwoast commented:
OK, we have the following new Init settings for jukeboxes, that just made sense alongside having a "shuffle button is invisible" setting.

shuffleButton: (true | false)  -- True by default, but may be set "false" to make the player appear without a shuffle button. By enabling it at default, I hope it alerts integrators to potential shuffle issues with how their jukebox is implemented.
shuffle: (true | false) -- Whether shuffle should be turned on at player load
startingTrack: (number OR "random") -- You can have a random startingTrack now, in conjunction with turning shuffle mode

Hopefully there are no crazy bugs hiding out ;) but testing is welcome and appreciated! :)

regosen commented:
Ok cool, I tried calling it like this and it works:

```
player = new Gapless5('player-block', {tracks: alltracks, playOnLoad: true, useHTML5Audio: false, shuffleButton: false});
```

That being said, would you mind adding a line about shuffleButton to the option comments?  Right now there's no mention of shuffleButton there:
```
// parameters are optional.  options:
//   tracks: path of file (or array of music file paths)
//   playOnLoad (default = false): play immediately
//   useWebAudio (default = true)
//   useHTML5Audio (default = false on mobile browsers, true otherwise)
```

Everything's almost working- the only thing remaining is that the playOnLoad option is still broken on the ZFP page.

wwoast commented:
Thankfully fixing playOnLoad was a one-variable fix. Updated comments, and added a couple notes to README.md too (though the README.md image should be updated with the shuffle button)

:) Fingers crossed this is finally ready!

regosen commented:
Great, that fixed it!

(I just noticed a bug with Safari while testing this, but I confirmed it happens on my version as well, so I'll open that as a separate issue on my end.)

Thanks for seeing this all the way to the end, I can tell you put a lot of time into it!  I'm merging this now.
